### PR TITLE
pca9685_pwm_out small improvement

### PR DESCRIPTION
**Pull request by SalimTerryLi**
**Describe problem solved by this pull request**
1. ~~reduce I2C calls~~
2. enable EXTCLK support
3. add support for schedule rate limit, to reduce I2C bus transfer but still maintain higher outputting resolution.
4. refactor to use state machine.

**Describe your solution**
Minor update.

### CODE REVIEW

File: src/drivers/pca9685_pwm_out/PCA9685.cpp [hunk 1]
```diff
@@ -221,59 +216,43 @@ void PCA9685::setDivider(uint8_t value)
...
+	int ret = transfer(buf, 2, nullptr, 0);
 
 	if (OK != ret) {
-		PX4_DEBUG("i2c::transfer returned %d", ret);
+		PX4_ERR("startOscillator: i2c::transfer returned %d", ret);
 		return;
 	}
 
-	usleep(5000);
+	usleep(500);
```
Comment (dagar) [inline src/drivers/pca9685_pwm_out/PCA9685.cpp:249]: This is likely fine, but ideally we wouldn't have any sleeps in bus threads.

File: src/drivers/pca9685_pwm_out/PCA9685.cpp [hunk 2]
```diff
@@ -79,14 +74,14 @@ int PCA9685::setFreq(float freq)
 	uint16_t realResolution = floorl((float)PCA9685_CLOCK_FREQ / freq);
 
 	if (realResolution < PCA9685_PWM_RES) { // unable to provide enough resolution
-		PX4_DEBUG("frequency too high");
+		PX4_ERR("frequency too high");
```
Comment (dagar) [inline src/drivers/pca9685_pwm_out/PCA9685.cpp:77]: Nitpick, this message probably isn't too useful without more context.

File: src/drivers/pca9685_pwm_out/PCA9685.cpp [hunk 3]
```diff
@@ -162,7 +156,7 @@ void PCA9685::setPWM(uint8_t channel, const uint16_t &value)
 	int ret = transfer(buf, 5, nullptr, 0);
 
 	if (OK != ret) {
-		PX4_DEBUG("i2c::transfer returned %d", ret);
+		PX4_ERR("setPWM: i2c::transfer returned %d", ret);
```
Comment (dagar) [inline src/drivers/pca9685_pwm_out/PCA9685.cpp:159]: Do you need these in general? They can be a bit dangerous if you hit a real persistent error and then start spamming the console. On a small embedded system trying to print to the console at high rate can disrupt the scheduling of other sensors. Maybe only an annoyance on a more powerful LInux system.

File: src/drivers/pca9685_pwm_out/main.cpp [hunk 4]
```diff
@@ -96,26 +108,29 @@ class PWMDriverWrapper : public cdev::CDev, public ModuleBase<PWMDriverWrapper>,
 
 	void updatePWMParamTrim();
 
+	int _schd_rate_limit;
```
Comment (dagar) [inline src/drivers/pca9685_pwm_out/main.cpp:111]: Initialize?

File: src/drivers/pca9685_pwm_out/PCA9685.cpp (hunk 2 above)
Comment (SalimTerryLi) [reply src/drivers/pca9685_pwm_out/PCA9685.cpp:77]: This is the whole output:
```
pxh> param set PWM_RATE 10
+ PWM_RATE: curr: 400 -> new: 10
pxh> ERROR [pca9685_pwm_out] frequency too low
ERROR [pca9685_pwm_out] failed to set pwm frequency, fall back to 50Hz
pxh> 

```

File: src/drivers/pca9685_pwm_out/PCA9685.cpp (hunk 3 above)
Comment (SalimTerryLi) [reply src/drivers/pca9685_pwm_out/PCA9685.cpp:159]: Thanks for pointing out that. I will keep only the ones from init.

File: src/drivers/pca9685_pwm_out/PCA9685.cpp (hunk 2 above)
Comment (dagar) [reply src/drivers/pca9685_pwm_out/PCA9685.cpp:77]: It's not a big deal, but I think a single message with context would be more useful, especially at bootup if you didn't just try to change `PWM_RATE`.

``` Console
ERROR [pca9685_pwm_out] failed to set PWM frequency to 10 Hz, fall back to 50Hz
```

Comment (SalimTerryLi) [reply src/drivers/pca9685_pwm_out/PCA9685.cpp]: Done.

**Comment by SalimTerryLi** [comment]
@dagar Hi there?

**Review by dagar — APPROVED** [review]

**Comment by mohanbabi** [comment]
Any wiki for how to use this feature

**Comment by SalimTerryLi** [comment]
@mohanbabi 

To enable external XTAL you can add definition

```
-DPCA9685_CLOCK_EXT=25000000.0
```

to the [board cmake file](https://github.com/PX4/PX4-Autopilot/blob/dbb0974b0cfb1582289b83684c656e90b051b698/boards/scumaker/pilotpi/default.cmake#L1-L3)

To set scheduling limit please refer to [-h flag](https://github.com/PX4/PX4-Autopilot/blob/dbb0974b0cfb1582289b83684c656e90b051b698/src/drivers/pca9685_pwm_out/main.cpp#L509-L544)

**Comment by mohanbabi** [comment]
@SalimTerryLi  Sir thank you.
 I am not that much of a progrmmer. in the bellow drone there used. And someone removed connections.  I am new to Px4. I need just user wiki for how to connections and configure
https://www.unmannedsystemstechnology.com/wp-content/uploads/2020/04/IF1000-Hexacopter.pdf

**Comment by SalimTerryLi** [comment]
> @SalimTerryLi Sir thank you.
> I am not that much of a progrmmer. in the bellow drone there used. And someone removed connections. I am new to Px4. I need just user wiki for how to connections and configure
> https://www.unmannedsystemstechnology.com/wp-content/uploads/2020/04/IF1000-Hexacopter.pdf

Sorry but you may not able to use this driver as it is not available on PixHawk platform for normal use case. Have a look here http://docs.px4.io/master/en/flight_controller/modalai_fc_v1.html and see if it helps

**Comment by mohanbabi** [comment]
@SalimTerryLi  thank you very much sir🙏

**Pull request merged**